src/nn/upscale.rs

```diff
@@ -0,0 +1,112 @@
+use crate::prelude::{ConstUpscale2D, NearestNeighbor, UpscaleMethod};
+#[cfg(feature = "nightly")]
+use crate::prelude::{Dtype, HasErr, Rank3, Rank4, Tape, Tensor, Upscale2DKernel, ZerosTensor};
+
+#[allow(unused)]
+use super::{BuildModule, Module, NonMutableModule, ZeroSizedModule};
+
+#[derive(Debug, Default, Clone)]
+pub struct Upscale2D<const OH: usize, const OW: usize = OH, M: UpscaleMethod = NearestNeighbor>(M);
+
+impl<const OH: usize, const OW: usize, M: UpscaleMethod> ZeroSizedModule for Upscale2D<OH, OW, M> {}
+impl<const OH: usize, const OW: usize, M: UpscaleMethod> NonMutableModule for Upscale2D<OH, OW, M> {}
+
+impl<const OH: usize, const OW: usize, M: UpscaleMethod, Img: ConstUpscale2D<M>> Module<Img>
+    for Upscale2D<OH, OW, M>
+{
+    type Output = Img::Output<OH, OW>;
+    type Error = Img::Err;
+
+    fn try_forward(&self, x: Img) -> Result<Self::Output, Img::Err> {
+        x.try_upscale2d()
+    }
+}
+
+#[cfg(feature = "nightly")]
+#[derive(Debug, Default, Clone)]
+pub struct Upscale2DBy<const H: usize, const W: usize = H, M: UpscaleMethod = NearestNeighbor>(M);
+
+#[cfg(feature = "nightly")]
+impl<const H: usize, const W: usize, M: UpscaleMethod> ZeroSizedModule for Upscale2DBy<H, W, M> {}
+#[cfg(feature = "nightly")]
+impl<const H: usize, const W: usize, M: UpscaleMethod> NonMutableModule for Upscale2DBy<H, W, M> {}
+
+#[cfg(feature = "nightly")]
+impl<
+        const H: usize,
+        const W: usize,
+        const IH: usize,
+        const IW: usize,
+        const C: usize,
+        E: Dtype,
+        M: UpscaleMethod,
+        D: Upscale2DKernel<E, M> + ZerosTensor<E>,
+        T: 'static + Tape<E, D>,
+    > Module<Tensor<Rank3<C, IH, IW>, E, D, T>> for Upscale2DBy<H, W, M>
+where
+    Tensor<Rank3<C, { IH * H }, { IW * W }>, E, D, T>: Sized,
+{
+    type Output = Tensor<Rank3<C, { IH * H }, { IW * W }>, E, D, T>;
+    type Error = <Self::Output as HasErr>::Err;
+
+    fn try_forward(
+        &self,
+        x: Tensor<Rank3<C, IH, IW>, E, D, T>,
+    ) -> Result<Self::Output, Self::Error> {
+        x.try_upscale2d()
+    }
+}
+
+#[cfg(feature = "nightly")]
+impl<
+        const H: usize,
+        const W: usize,
+        const IH: usize,
+        const IW: usize,
+        const B: usize,
+        const C: usize,
```

Comment:
We can probably make these generic dims
```suggestion
        B: Dim,
        C: Dim,
```